pkg/daemon/controller/controller.go

```diff
@@ -265,6 +276,86 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
+	err = wait.PollUntilContextTimeout(sc.ctx, 15*time.Second, 5*time.Minute, true,
+		func(ctx context.Context) (done bool, err error) {
+			stillNeedsReboot := false
+			for _, varName := range cluster.SupportedProxyVars {
+				cmd := fmt.Sprintf("[Environment]::GetEnvironmentVariable('%s', 'Process')", varName)
+				out, err := sc.psCmdRunner.Run(cmd)
+				if err != nil {
+					return false, fmt.Errorf("error running PowerShell command %s with output %s: %w", cmd, out, err)
+				}
+				if strings.TrimSpace(out) != envVars[varName] {
+					stillNeedsReboot = true
+				}
+			}
+			return !stillNeedsReboot, nil
+		})
+	if err != nil {
+		return true, fmt.Errorf("error waiting for environment vars to get picked up by processes: %w", err)
+	}
+	// Remove the reboot annotation after we know the reboot occurred successfully. No-op if already not present
+	return false, metadata.RemoveRebootAnnotation(sc.ctx, sc.client, node)
+}
+
+// ensureVarsAreUpToDate ensures that the proxy environment variables are set as expected on the instance
+// If there's any changes, an instance restart is required to ensure all processes pick up the updated values.
+func (sc *ServiceController) ensureVarsAreUpToDate(envVars map[string]string) (restartRequired bool, err error) {
+	// systemEnvVarRegistryPath is where system level environment variables are stored in the Windows OS
+	const systemEnvVarRegistryPath = `SYSTEM\CurrentControlSet\Control\Session Manager\Environment`
+	restartRequired = false
+	for key, expectedVal := range envVars {
+		registryKey, err := registry.OpenKey(registry.LOCAL_MACHINE, systemEnvVarRegistryPath, registry.ALL_ACCESS)
+		if err != nil {
+			return false, fmt.Errorf("unable to open Windows system registry key %s: %w", systemEnvVarRegistryPath, err)
+		}
+		defer func() { // always close the registry key, without swallowing any error returned before the defer call
+			closeErr := registryKey.Close()
+			if closeErr != nil {
+				if err != nil {
+					err = fmt.Errorf("multiple errors: %w, %w", err, closeErr)
+				}
+				err = closeErr
+			}
+		}()
+
+		actualVal, _, err := registryKey.GetStringValue(key)
+		if err != nil && err != registry.ErrNotExist {
+			return false, fmt.Errorf("unable to read environment variable %s: %w", key, err)
+		}
+
+		if actualVal != expectedVal {
+			klog.Infof("updating environment variable %s", key)
+			// Because we modify env vars are the "system" level rather than the ephemeral "process" level,
+			// we cannot use os.Setenv, which is a wrapper for syscall.SetEnvironmentVariable
+			// As per Microsoft docs: "Calling SetEnvironmentVariable has no effect on the system environment variables"
+			err := registryKey.SetStringValue(key, expectedVal)
+			if err != nil {
+				// Do not log value as proxy information is sensitive
+				return false, fmt.Errorf("unable to set environment variable %s: %w", key, err)
+			}
+			restartRequired = true
+		}
+	}
+	return restartRequired, nil
```

Comment:
Shouldn't this be `return restartRequired, err`?

Reply:
The err value should be `nil` as this is the happy path. By the time we get here, if there was a non-nil error of note, we'd have returned already

Reply:
@saifshaikh48 we only set the error in the defer function but where do we return it, if we don't return it here?

Reply:
When the overarching function (ensureVarsAreUpToDate) returns, its second return value is stored in the variable `err`. Then the defer function can access and modify it. And finally, after the defer function runs, whatever value is in `err` is returned

Reply:
Did you confirm this in a go playground?

Reply:
yes, https://go.dev/play/p/JTCvFeEsYrs